Sources/SwiftBasicFormat/InferIndentation.swift

```diff
@@ -0,0 +1,127 @@
+  ///
+  /// We start with .newline to indicate that the first token starts on a newline, even if it technically doesn't have
+  /// a leading newline character.
+  private var previousTokenTrailingTrivia: Trivia = .newline
+
+  /// Counts how many lines had how many spaces of indentation.
+  ///
+  /// For example, spaceIndentedLines[2] = 4 means that for lines had exactly 2 spaces of indentation.
+  private var spaceIndentedLines: [Int: Int] = [:]
+
+  /// See `spaceIndentedLines`
+  private var tabIndentedLines: [Int: Int] = [:]
+
+  /// The number of lines that were processed for indentation inference.
+  ///
+  /// This will be lower than the actual number of lines in the syntax node because
+  ///  - It does not count lines without indentation
+  //// - It does not count newlines in block doc comments (because we don't process the comment's contents)
+  private var linesProcessed = 0
+
+  override func visit(_ token: TokenSyntax) -> SyntaxVisitorContinueKind {
+    defer { previousTokenTrailingTrivia = token.trailingTrivia }
+    let triviaAtStartOfLine =
+      (previousTokenTrailingTrivia + token.leadingTrivia)
+      .drop(while: { !$0.isNewline })  // Ignore any trivia that's on the previous line
+      .split(omittingEmptySubsequences: false, whereSeparator: \.isNewline)  // Split trivia into the lines it occurs on
+      .dropFirst()  // Drop the first empty array; exists because we dropped non-newline prefix and newline is separator
+
+    LINE_TRIVIA_LOOP: for lineTrivia in triviaAtStartOfLine {
+      switch lineTrivia.first {
+      case .spaces(var spaces):
+        linesProcessed += 1
+        for triviaPiece in lineTrivia.dropFirst() {
+          switch triviaPiece {
+          case .spaces(let followupSpaces): spaces += followupSpaces
+          case .tabs: break LINE_TRIVIA_LOOP  // Count as processed line but don't add to any indentation count
+          default: break
+          }
+        }
+        spaceIndentedLines[spaces, default: 0] += 1
+      case .tabs(var tabs):
+        linesProcessed += 1
+        for triviaPiece in lineTrivia.dropFirst() {
+          switch triviaPiece {
+          case .tabs(let followupTabs): tabs += followupTabs
+          case .spaces: break LINE_TRIVIA_LOOP  // Count as processed line but don't add to any indentation count
+          default: break
+          }
+        }
+        tabIndentedLines[tabs, default: 0] += 1
+      default:
+        break
+      }
+    }
+    return .skipChildren
+  }
+
+  static func inferIndentation(of tree: some SyntaxProtocol) -> Trivia? {
+    let visitor = IndentationInferrer(viewMode: .sourceAccurate)
+    visitor.walk(tree)
+    if visitor.linesProcessed < 3 {
+      // We don't have enough lines to infer indentation reliably
+      return nil
+    }
+
+    // Pick biggest indentation that encompasses at least 90% of the source lines.
+    let threshold = Int(Double(visitor.linesProcessed) * 0.9)
+
+    for spaceIndentation in [8, 4, 2] {
```

Comment:
Poor 3 space users 😅. Is 8 actually a thing that people use?

I think one issue we're likely to run into here is that for highly nested snippets, this is often going to return 8 even if it's eg. 2 space. Is the idea that this would only be run on entire files/functions/something else?

Reply:
That shouldn’t be an issue. If we have 4 lines indented with 8 spaces and 1 line indented with 10 spaces, we’ll infer the indentation to 2 spaces because only 80% of the lines have an indentation that’s a multiple of 8 but all lines have an indentation that’s a multiple of 2. 

And I think I’ve seen codebases with 8 spaces of indentation. Can’t remember where and when, though.